### ISSUE TITLE
pretrain slide encoder

### ISSUE THREAD

**Issue by linuxbioinfoguy** [prov-gigapath/prov-gigapath#77]
Great work on this study and it may provide a lot of example for the later on researches. My question is that how do you pretrain the slide encoder (by the longnet), from the repository it seems there is seems no code for how to train the slide encoder, is it possible to have some example?  assuming it does not need any slide level label, correct?  

Thanks!

**Comment by HanwenXuTHU** [comment]
Hi, we are mainly using the masked image modelling approaches to pretrain the slide encoder. I adapted the official [MAE](https://github.com/facebookresearch/mae)'s codebase by replacing the ViT there with GigaPath's slide encoder. It wouldn't be too hard to adapt this codebase. Let me know if you encounter any issues when adapting MAE. Would love to help!

**Comment by linuxbioinfoguy** [comment]
Hi, Xu, thanks very much for the reply, I appreciated that. I briefly look at the MAE repository, correct me if I am wrong below: if you referring using the slide encoder to replace the model in the file model_vit, it sounds like you only use their distribution training as well as datafeed that kind of infrastructure, is that correct; for us we already have our own data loader (for the tile embedding and the position as described in the repository), could I assume that using the code in the your demo 4 (load_slide_encoder), instead of load the model, create a model with random initialization and put the mode to train, then feed the data (tile feature and coords), that would serve the same pre-train purpose? thanks!

**Comment by HanwenXuTHU** [comment]
That's what I did when I pretrained the slide encoder: masked some portions of tile embeddings and use MAE to reconstruct. I also used their masking strategy: masking some portions of the tile embeddings and learn to reconstruct. I would love to take a look at your code if you encounter any problem when implementing it.

**Comment by linuxbioinfoguy** [comment]
Thanks Xu, my code is fairly simple as mentioned above, I assumed just create a longnet model as in your demo_4 with random initialization, then I already have the features/coords loaded (from dinov2), load that directly to the model as follow:                                                                                                                                                                                                      
     model = slide_encoder.create_model("", gigapath_slide_enc12l768d, 1536) ; then after read the tiles/coords, feed to the model as :            model(
                        x = torch.tensor(
                            embedding_batch[
                                k * cfg["train"]["batch_size_per_gpu"]: (k + 1) * cfg["train"]["batch_size_per_gpu"]
                            ]
                        ).to(gpudevice),

                        coords = torch.tensor(
                            coords_batch[
                                k * cfg["train"]["batch_size_per_gpu"]: (k + 1) * cfg["train"]["batch_size_per_gpu"]
                            ]
                        ).to(gpudevice),
                    )  ,  so based on your comment seems that was not right, it would be great for a little more detail (if there is a repository on this pretrain slide encoder code, that would be great!  thanks!

**Comment by linuxbioinfoguy** [comment]
would it be possible for you to share the code to replace the forward_decoder function in the mae repository? thanks!

**Comment by adas2125** [comment]
Hi @linuxbioinfoguy, I hope you are doing well. I am also interested in the pre training of the slide encoder and wanted to ask if you are utilizing the `models_vit.py` script or the `model_mae.py` script for your pre-training tasks. From my understanding, the `models_vit.py` script is primarily intended for fine-tuning, but I wanted to clarify your approach.

Additionally, regarding your question about `forward_decoder`, it seems that some of the scripts within the `gigapath/torchscale/architecture/` directory could potentially be adapted to address this. I plan to explore this further as well, so I would appreciate it if you could share any progress or insights you make along the way.

Thanks.

**Comment by linuxbioinfoguy** [comment]
Hi, @adas2125, thanks for the comment, the model_mae.py is basically the classification head for the fine-tuning, so it is useless in this case. I also saw the Encoder/Decoder definition inside the torchscale. In order to utilize those code, a lot of new code or revision need to be add in into a new module that by using models_mae.py. meanwhile, it would be great that the gigapath folks could release their training parameter in detail but not just like "use the default setting" for mae and specify how they evaluate the authenticity of the model. of course it can be say that based on the loss however, the loss function in the mae repository is fairly simple, I wonder if more sophisticated one was used. I am not sure if you or the gigapath folks notice that too, I happen to load the gigapath slide model without actually load the model parameter, so basically it is a empty model with random initialization, however, in a binary classification, the final output of 768(full output is over 9000 features) for predicting binary classification (like the mutation, just for example, I used in my own classification purpose), you will find such empty slide level model, actually have similar or even better performance (by using the generated slide level embeding for classification), and this was repeated many many times and it is a systematic issue using applying too many feature into a too small data set for validation, so if the classification validation set in the paper is more that couple 10s of thousands, nevermind, but if it is smaller, though such systematic issue is hard to evaluate/comment, it indeed lower the confidence of using the "trained" model at any checkpoint.  and some of the parameter/strategy in the MAE, does worry me a little bit, for example the default is 75% masking, for the imagenet that kind of data, making that high (like dog/cat something) may be ok as those have a clear and predictable shape, however, in the AP pathology it is completely different, there is no clear "shape" of tumor of any kind, especially a lot of diffusion type. so masking that high percentage worries me like, for example, most of the tumor got masked, and the model trying to regenerate the tumor tissue. this may lead to a input without tumor at all but the model trying to makeup something on it, which may not be logically sound.
